dokka-runners/dokka-gradle-plugin/src/testFixtures/kotlin/kotestFiles.kt

```diff
@@ -3,106 +3,140 @@
+    }
+    if (onlyInActual.isNotEmpty()) {
+        appendLine("actualDir has ${onlyInActual.size} unexpected files:")
+        appendLine(onlyInActual.sorted().joinToFormattedList())
     }
-}
 
+    // Compare contents of files that are present in both directories
+    val commonFiles = actualFiles intersect expectedFiles
 
-private fun File.shouldHaveSameStructureAndContentAs2(
-    file: File,
-    filterLhs: (File) -> Boolean = { false },
-    filterRhs: (File) -> Boolean = { false },
-    fileAssert: FileAsserter,
-) {
-    val expectFiles = this.walkTopDown().filter(filterLhs).toList()
-    val actualFiles = file.walkTopDown().filter(filterRhs).toList()
-
-    expectFiles shouldBeSameSizeAs actualFiles
-
-    val assertContext = FileAsserter.Context(
-        expectParentPath = this.invariantSeparatorsPath,
-        actualParentPath = file.invariantSeparatorsPath,
-    )
-
-    expectFiles.zip(actualFiles) { expect, actual ->
-        when {
-            expect.isDirectory -> actual.shouldBeADirectory()
-            expect.isFile -> {
-                with(fileAssert) {
-                    assertContext.assert(expect, actual)
-                }
-            }
+    commonFiles
+        .sorted()
+        .forEach { relativePath ->
+            val expectedFile = expectedDir.resolve(relativePath)
+            val actualFile = actualDir.resolve(relativePath)
+
+            val expectedLines = expectedFile.readByteLines()
+            val actualLines = actualFile.readByteLines()
+
+            val patch = DiffUtils.diff(expectedLines, actualLines)
 
-            else -> error("There is an unexpected error analyzing file trees. Failed to determine filetype of $expect")
+            if (patch.deltas.isNotEmpty()) {
+                appendLine("${relativePath.invariantSeparatorsPathString} has ${patch.deltas.size} differences in content:")
+
+                val diff = UnifiedDiffUtils.generateUnifiedDiff(
+                    /* originalFileName = */ expectedFile.relativeTo(expectedDir).invariantSeparatorsPathString,
+                    /* revisedFileName = */ actualFile.relativeTo(actualDir).invariantSeparatorsPathString,
+                    /* originalLines = */ expectedLines,
+                    /* patch = */ patch,
+                    /* contextSize = */ 3,
+                )
+
+                appendLine(diff.joinToString("\n").prependIndent())
+            }
         }
-    }
 }
 
 
-private fun interface FileAsserter {
+/**
+ * Pretty print files as a list.
+ */
+private fun Collection<Path>.joinToFormattedList(limit: Int = 10): String =
+    joinToString("\n", limit = limit) { "  - ${it.invariantSeparatorsPathString}" }
+
 
-    data class Context(
-        val expectParentPath: String,
-        val actualParentPath: String,
-    )
+/**
+ * Read lines from a file, leniently.
+ * Handles text and binary data (though binary data might be hard to read).
+ *
+ * ([kotlin.io.path.readLines] blows up when it reads binary files.)
```

Comment:
as I mentioned before, IMO, I think for binary files it will be better to just check size/hash and not content